### PR TITLE
Move remove duplicates logic to CSVM module

### LINKED ISSUE #178: Move Subset ids creation to CSVM logic
In a GaussianMixture execution, we have detected that much time is lost creating the element _ids in the Subset constructor. The same could happen for other algorithms. These ids are only used in CSVM for removing duplicates when concatenating subsets. So we have to move this logic to the CSVM module, to remove the impact on other algorithms.

### PR DESCRIPTION
**Pull request by salvisolamartinell**
# Description

Move the logic that creates ids for each element in a Subset and uses them for the optional remove_duplicates to the CSVM module. Test for Subset removed.

Fixes #178

## Type of change

Please delete options that are not relevant.

- [ ] New algorithm or support class.
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. If you haven't added any test and it is relevant provide instructions so we can reproduce.

- [ ] I have added a new test file: [E.g. `test_rf.py`]
- [ ] I have added a new test case: [E.g. `RFTest.test_make_classification`]
- [x] I have tested it manually in a **local environment**.
- [ ] I have tested it manually in a **supercomputer**.

Reproduce instructions:
I checked that the current tests cover the 'remove duplicates' logic.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have documented the public methods of any public class according to the guide styles. 
- [ ] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes
- [x] I have rebased my branch before trying to merge.

### CODE REVIEW

File: dislib/classification/csvm/base.py [hunk 1]
```diff
@@ -355,10 +384,38 @@ def _merge_scores(*partials):
...
+    return Subset(samples, labels), ids
 
-    return set0
+
+def _merge_pair(set_0, set_1):
+    samples_0, labels_0, ids_0 = set_0
+    samples_1, labels_1, ids_1 = set_1
+    assert issparse(samples_0) == issparse(samples_1), \
+        "Cannot concatenate sparse data with non-sparse data."
+    assert (labels_0 is None) == (labels_1 is None), \
```
Comment (javicid) [inline dislib/classification/csvm/base.py]: These checks are unnecessary as you can assume that all subsets are either dense or sparse.

Also, you can assume that all subsets have labels.

File: dislib/classification/csvm/base.py [hunk 2]
```diff
@@ -355,10 +384,38 @@ def _merge_scores(*partials):
...
+        labels = np.concatenate((labels_0, labels_1))
+    ids = np.concatenate((ids_0, ids_1))
+    ids, uniques = np.unique(ids, return_index=True)
+    indices = np.argsort(uniques)
+    uniques = uniques[indices]
+    ids = ids[indices]
+    samples = samples[uniques]
+    if labels is not None:
+        labels = labels[uniques]
+    return samples, labels, ids
```
Comment (javicid) [inline dislib/classification/csvm/base.py]: To generate the new Subset (without duplicates), you can use Subset.concatenate and then slice it using subset[uniques]

Comment (salvisolamartinell) [reply dislib/classification/csvm/base.py]: Absolutely right.

File: dislib/classification/csvm/base.py (hunk 1 above)
Comment (salvisolamartinell) [reply dislib/classification/csvm/base.py]: Okay. Working on it.

File: dislib/classification/csvm/base.py (hunk 2 above)
Comment (salvisolamartinell) [reply dislib/classification/csvm/base.py]: Done.

File: dislib/classification/csvm/base.py (hunk 1 above)
Comment (salvisolamartinell) [reply dislib/classification/csvm/base.py]: Done.

**Comment by salvisolamartinell** [comment]
@javicid  I did:

> 1. Place the creation of the ids in separate tasks at the beginning (before training).
and didn't do 
> 2. Use tuples of (subset, ids) as future objects instead of working with two separate lists.
because, as commented, this would imply reserializing the input subsets together with their ids, or complicating the code much more.

**Pull request merged**